File: LeetCode/ArrayStrings/python/longestCommonPrefix.py
from typing import List


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        strs.sort()
        n = len(strs)
        s1 = strs[0]
        s2 = strs[n-1]
        res = ""
        i = 0

        while i < len(s1) and s1[i] == s2[i]:
            res += s1[i]
            i += 1

        return res

    @staticmethod
    def testSolution(input) -> None:
        print("Input: strs: {}".format(input["strs"]))
        print("Expected: {}".format(input["expected"]))
        print("Result: {}".format(Solution().longestCommonPrefix(input["strs"])))
        print("-" * 50)

if __name__ == "__main__":

    input = {"strs": ["flower","flow","flight"], "expected": "fl"}
    Solution.testSolution(input)

    input = {"strs": ["dog","racecar","car"], "expected": ""}
    Solution.testSolution(input)

File: LeetCode/ArrayStrings/java/longestCommonPrefix.java
package LeetCode.ArrayStrings.java;

import java.util.Arrays;

record longestCommonPrefixRecord(String[] strs, String expected) {
}

public class longestCommonPrefix {
    public String longestCommonPrefixSolution(String[] strs) {

        // We need to sort the words in the strings
        // After it's sorted, we compare the words at index 0 and the last index chars
        // by chars
        // Exit if we see that there's no common chars
        Arrays.sort(strs);

        int n = strs.length;

        String s1 = strs[0];
        String s2 = strs[n - 1];

        StringBuilder res = new StringBuilder();

        int i = 0;

        // Because this is sorted, s1 will contain the string with smaller length
        while (i < s1.length() && s1.charAt(i) == s2.charAt(i)) {
            res.append(String.valueOf(s1.charAt(i)));
            i += 1;
        }

        return res.toString();
    }

    public static void main(String[] args) {
        longestCommonPrefixRecord input = new longestCommonPrefixRecord(
                new String[] { "flower", "flow", "flight" },
                "fl");

        longestCommonPrefix.testSolution(input);

        input = new longestCommonPrefixRecord(
                new String[] { "dog", "racecar", "car" },
                "");

        longestCommonPrefix.testSolution(input);
    }

    private static void testSolution(longestCommonPrefixRecord input) {
        System.out.println("Input: strs: " + Arrays.toString(input.strs()));
        System.out.println("Expected: " + input.expected());
        System.out.println("Result: " + new longestCommonPrefix().longestCommonPrefixSolution(input.strs()));
        System.out.println("-".repeat(50));
    }

}
